jets/c/rip.c

```diff
@@ -1,92 +1,81 @@
-/* j/3/rip.c
-**
-*/
 #include "all.h"
 
+u3_noun u3qc_rip(u3_atom bloq, u3_atom b) {
+  if ( !_(u3a_is_cat(bloq)) || (bloq >= 32) ) {
+    return u3m_bail(c3__fail);
+  }
 
-/* functions
-*/
-  u3_noun
-  u3qc_rip(u3_atom a,
-           u3_atom b)
-  {
-    if ( !_(u3a_is_cat(a)) || (a >= 32) ) {
-      return u3m_bail(c3__fail);
-    }
-    else {
-      u3_noun pir = u3_nul;
-      c3_g a_g = a;
-      c3_w i_w;
-
-      if ( a_g < 5 ) {
-        c3_w met_w = u3r_met(a_g, b);
-        c3_w mek_w = ((1 << (1 << a_g)) - 1);
-
-        for ( i_w = 0; i_w < met_w; i_w++ ) {
-          c3_w pat_w = (met_w - (i_w + 1));
-          c3_w bit_w = (pat_w << a_g);
-          c3_w wor_w = (bit_w >> 5);
-          c3_w sif_w = (bit_w & 31);
-          c3_w src_w = u3r_word(wor_w, b);
-          c3_w rip_w = ((src_w >> sif_w) & mek_w);
-
-          pir = u3nc(rip_w, pir);
-        }
-        return pir;
-      }
-      else {
-        c3_w met_w = u3r_met(a_g, b);
-        c3_w len_w = u3r_met(5, b);
-        c3_g san_g = (a_g - 5);
-        c3_w san_w = 1 << san_g;
-        c3_w dif_w = (met_w << san_g) - len_w;
-        c3_w tub_w = ((dif_w == 0) ? san_w : (san_w - dif_w));
-
-        for ( i_w = 0; i_w < met_w; i_w++ ) {
-          c3_w  pat_w = (met_w - (i_w + 1));
-          c3_w  wut_w = (pat_w << san_g);
-          c3_w  sap_w = ((0 == i_w) ? tub_w : san_w);
-          c3_w* sal_w = u3a_slab(sap_w);
-
-          if ( 0 == sal_w ) {
-            return u3m_bail(c3__fail);
-          } else {
-            c3_w    j_w;
-            u3_atom rip;
-
-            for ( j_w = 0; j_w < sap_w; j_w++ ) {
-              sal_w[j_w] = u3r_word(wut_w + j_w, b);
-            }
-
-            rip = u3a_malt(sal_w);
-            pir = u3nc(rip, pir);
-          }
-          len_w -= san_w;
-        }
-      }
-      return pir;
+  c3_g bloq_g = bloq;
+
+  if ( bloq_g < 5 ) {                                   //  produce direct atoms
+    u3_noun acc   = u3_nul;
+    c3_w    met_w = u3r_met(bloq_g, b);                 //  num blocks in atom
+    c3_w    nbits = 1 << bloq_g;                        //  block size in bits
+    c3_w    bmask = ((1 << nbits) - 1);                 //  result mask
+
+    for ( c3_w i_w = 0; i_w < met_w; i_w++ ) {          //  `i_w` is block index
+      c3_w nex_w = i_w + 1;                             //  next block
+      c3_w pat_w = met_w - nex_w;                       //  blks left after this
+      c3_w bit_w = pat_w << bloq_g;                     //  bits left after this
+      c3_w wor_w = bit_w >> 5;                          //  wrds left after this
+      c3_w sif_w = bit_w & 31;                          //  bits left in word
+      c3_w src_w = u3r_word(wor_w, b);                  //  find word by index
+      c3_w rip_w = ((src_w >> sif_w) & bmask);          //  get item from word
+
+      acc = u3nc(rip_w, acc);
```

Comment:
If `rip_w` has its high bit set, it will become a pointer to somewhere else on the loom! I think we should either add an assert or a function that converts a `c3_w` to a noun explicitly.

Reply:
Or ... `bloq_g < 5` means the max bloq size is 2^4, or 16 bits. If that's the case, what are we doing with, say, 17-bit atoms? I'm not sure it's valid to have a 17-bit indirect atom.

I kinda feel like we should just be calling whatever numberToAtom() function we have on each slice and not think about this so hard.

Reply:
This whole block looks like a performance optimization for small bloq sizes. Since that likely matches most calls to `+rip`, it's probably worth doing this way.

Reply:
This is the same code as before, it just adds comments.

Reply:
Yeah, that's right @joemfb. The case where we are producing direct atoms should be *way* faster

@belisarius222 `bloq_g` needs to be less than 5 for exactly the reason you pointed out if it was equal to five, then we might accidentally produce direct atoms with the high bit set, but since it isn't, we don't have to worry about it because `bmask` wont allow that.